In [ ]:
import pandas as pd

In [ ]:
# data = pd.read_table('/content/chipotle.tsv')

In [ ]:
data = pd.read_csv('/content/chipotle.tsv', sep='\t')

In [ ]:
data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

In [ ]:
data.tail()

order_id  quantity           item_name  \
4617      1833         1       Steak Burrito   
4618      1833         1       Steak Burrito   
4619      1834         1  Chicken Salad Bowl   
4620      1834         1  Chicken Salad Bowl   
4621      1834         1  Chicken Salad Bowl   

                                     choice_description item_price  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75

In [ ]:
data.shape

(4622, 5)

In [ ]:
data.columns

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


In [ ]:
data.describe()

order_id     quantity
count  4622.000000  4622.000000
mean    927.254868     1.075725
std     528.890796     0.410186
min       1.000000     1.000000
25%     477.250000     1.000000
50%     926.000000     1.000000
75%    1393.000000     1.000000
max    1834.000000    15.000000

# Missing / NULL Values

In [ ]:
data.isnull().sum()

order_id                 0
quantity                 0
item_name                0
choice_description    1246
item_price               0
dtype: int64

In [ ]:
data[data['choice_description'].isnull()]

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
6            3         1                          Side of Chips   
10           5         1                    Chips and Guacamole   
14           7         1                    Chips and Guacamole   
...        ...       ...                                    ...   
4600      1827         1                    Chips and Guacamole   
4605      1828         1                    Chips and Guacamole   
4613      1831         1                                  Chips   
4614      1831         1                          Bottled Water   
4616      1832         1                    Chips and Guacamole   

     choice_description item_price  
0                   NaN     $2.39   
3                   NaN     $2.39   
6                   NaN     $1.69   
10                  NaN     $4.45   
14                  NaN     $4.45   
...                 ...        ...  
4600                NaN     $4.45   
4605                NaN     $4.45   
4613                NaN     $2.15   
4614                NaN     $1.50   
4616                NaN     $4.45   

[1246 rows x 5 columns]

As NULL values are large in number we cannot that many number of values from the dataframe. We can instead replace NULL values with default values or Imputing based on Other Columns.

In [ ]:
# Replacing with default value
data['choice_description'] = data['choice_description'].fillna('Not Specified')

In [ ]:
data.isnull().sum()

order_id              0
quantity              0
item_name             0
choice_description    0
item_price            0
dtype: int64

# Data Types

In [ ]:
data.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

# Duplicated Entries


In [ ]:
data[data.duplicated()]

order_id  quantity                            item_name  \
238        103         1                        Steak Burrito   
248        108         1                          Canned Soda   
297        129         1                        Steak Burrito   
381        165         1                    Canned Soft Drink   
484        205         1                         Chicken Bowl   
567        233         1                    Canned Soft Drink   
617        254         1                                Chips   
684        282         1                    Canned Soft Drink   
773        319         1                         Chicken Bowl   
908        376         1                        Steak Burrito   
924        381         1                   Chicken Soft Tacos   
1003       412         1                        Bottled Water   
1117       462         1                      Chicken Burrito   
1203       491         1                      Chicken Burrito   
1245       507         1                    Canned Soft Drink   
1381       561         1                    Canned Soft Drink   
1410       573         1                        Bottled Water   
1420       576         1                        Barbacoa Bowl   
1421       576         1                  Barbacoa Salad Bowl   
1502       610         1                      Chicken Burrito   
1534       624         1  Chips and Tomatillo-Red Chili Salsa   
1548       628         1                  Chips and Guacamole   
1612       651         1                                 Izze   
1791       724         1                    Canned Soft Drink   
1792       724         1                    Canned Soft Drink   
2078       838         1                   Chicken Soft Tacos   
2127       856         1                      Chicken Burrito   
2307       926         1                         Chicken Bowl   
2318       926         1                         Chicken Bowl   
2362       940         1                    Canned Soft Drink   
2448       972         1                         Chicken Bowl   
2553      1011         1                    Canned Soft Drink   
2554      1011         1                    Canned Soft Drink   
2560      1013         1                  Chips and Guacamole   
2576      1020         1  Chips and Tomatillo Red Chili Salsa   
2662      1057         1                   Chicken Soft Tacos   
2708      1074         1                                Chips   
2717      1078         1                         Chicken Bowl   
2833      1125         1                      Chicken Burrito   
2881      1144         1                        Bottled Water   
2973      1182         1                         Chicken Bowl   
2977      1184         1                           Steak Bowl   
3121      1244         1                  Chips and Guacamole   
3430      1377         1  Chips and Tomatillo Red Chili Salsa   
3637      1454         1                   Chicken Soft Tacos   
3716      1484         1                          Canned Soda   
3764      1507         1                      Chicken Burrito   
3829      1534         1                      Chicken Burrito   
3924      1574         1                        Bottled Water   
4112      1647         1                         Chicken Bowl   
4155      1660         1                  Chips and Guacamole   
4236      1691         1                         Chicken Bowl   
4266      1701         1                                Chips   
4372      1745         1                  Chips and Guacamole   
4463      1778         1                    Canned Soft Drink   
4479      1785         1                        Bottled Water   
4544      1806         1                        Bottled Water   
4570      1817         1                        Bottled Water   
4603      1827         1                     Barbacoa Burrito   

                                     choice_description item_price  
238   [Tomatillo Red Chili Salsa, [R

# Quantity and Item Price

In [ ]:
data[['quantity','item_name','item_price']]

quantity                              item_name item_price
0            1           Chips and Fresh Tomato Salsa     $2.39 
1            1                                   Izze     $3.39 
2            1                       Nantucket Nectar     $3.39 
3            1  Chips and Tomatillo-Green Chili Salsa     $2.39 
4            2                           Chicken Bowl    $16.98 
...        ...                                    ...        ...
4617         1                          Steak Burrito    $11.75 
4618         1                          Steak Burrito    $11.75 
4619         1                     Chicken Salad Bowl    $11.25 
4620         1                     Chicken Salad Bowl     $8.75 
4621         1                     Chicken Salad Bowl     $8.75 

[4622 rows x 3 columns]

In [ ]:
data[data['quantity'] < 0]

Empty DataFrame
Columns: [order_id, quantity, item_name, choice_description, item_price]
Index: []

In [ ]:
print(data['quantity'].min())
print(data['quantity'].max())

1
15


# Handling Special Characters

# Choice Description

In [ ]:
data['choice_description'].head(10)

0                                        Not Specified
1                                         [Clementine]
2                                              [Apple]
3                                        Not Specified
4    [Tomatillo-Red Chili Salsa (Hot), [Black Beans...
5    [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...
6                                        Not Specified
7    [Tomatillo Red Chili Salsa, [Fajita Vegetables...
8    [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...
9    [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...
Name: choice_description, dtype: object

In [ ]:
data.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

In [ ]:
first_choice_description = data['choice_description'][4]
print(first_choice_description)

[Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]


In [ ]:
def process_choice(choice):
    result = []
    if isinstance(choice, list):
        for item in choice:
            if isinstance(item, list):
                result.append(', '.join(str(inner_item) for inner_item in item))
            else:
                result.append(str(item))
    else:
        result.append(str(choice))
    return result

choice_description =['Tomatillo-Red Chili Salsa (Hot)', ['Black Beans', 'Rice', 'Cheese', 'Sour Cream']]

processed_items = process_choice(choice_description)
print(f"Processed Items: {processed_items}")

Processed Items: ['Tomatillo-Red Chili Salsa (Hot)', 'Black Beans, Rice, Cheese, Sour Cream']


we can add extract each elements from the list and can create new column and just store the names in it and we can also create an additional column which can contain total no. of elements in the choice_description column

# item_price

In [ ]:
data['item_price'] = data['item_price'].apply(lambda x: x.replace("$",""))

In [ ]:
data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                      Not Specified      2.39   
1                                       [Clementine]      3.39   
2                                            [Apple]      3.39   
3                                      Not Specified      2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...     16.98

# Order ID

In [ ]:
data['order_id'].duplicated().sum()

2788

**order_id** column contains total 2788 duplicate order_id

# Standardize item names

In [ ]:
data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                      Not Specified      2.39   
1                                       [Clementine]      3.39   
2                                            [Apple]      3.39   
3                                      Not Specified      2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...     16.98

In [ ]:
# Function to standardize item names
def standardize_item_name(item_name):
    standardized_name = item_name.lower().replace(' and ', ', ').replace('-', ' ')
    return standardized_name

# data['item_name_standardized'] = data['item_name'].apply(standardize_item_name)
data['item_name'] = data['item_name'].apply(standardize_item_name)

In [ ]:
data.head()

order_id  quantity                           item_name  \
0         1         1           chips, fresh tomato salsa   
1         1         1                                izze   
2         1         1                    nantucket nectar   
3         1         1  chips, tomatillo green chili salsa   
4         2         2                        chicken bowl   

                                  choice_description item_price  
0                                      Not Specified      2.39   
1                                       [Clementine]      3.39   
2                                            [Apple]      3.39   
3                                      Not Specified      2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...     16.98

# Quantity and Price Relationships

In [ ]:
data.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

In [ ]:
data['item_price'] = data['item_price'].astype(float)

In [ ]:
data.dtypes

order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price            float64
dtype: object

In [ ]:
# Calculate total price for each item
data['total_price'] = data['quantity'] * data['item_price']

data[['quantity', 'item_price', 'total_price']]

quantity  item_price  total_price
0            1        2.39         2.39
1            1        3.39         3.39
2            1        3.39         3.39
3            1        2.39         2.39
4            2       16.98        33.96
...        ...         ...          ...
4617         1       11.75        11.75
4618         1       11.75        11.75
4619         1       11.25        11.25
4620         1        8.75         8.75
4621         1        8.75         8.75

[4622 rows x 3 columns]

In [ ]:
data.dtypes

order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price            float64
total_price           float64
dtype: object

# Convert to csv

In [ ]:
print('\n Cleaned dataset')
data.head()


 Cleaned dataset


order_id  quantity                           item_name  \
0         1         1           chips, fresh tomato salsa   
1         1         1                                izze   
2         1         1                    nantucket nectar   
3         1         1  chips, tomatillo green chili salsa   
4         2         2                        chicken bowl   

                                  choice_description  item_price  total_price  
0                                      Not Specified        2.39         2.39  
1                                       [Clementine]        3.39         3.39  
2                                            [Apple]        3.39         3.39  
3                                      Not Specified        2.39         2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98        33.96

In [ ]:
data.to_csv('cleaned_dataset.csv', sep=',', index=False)

In [ ]:
x = pd.read_csv('cleaned_dataset.csv')

In [ ]:
x.head()

order_id  quantity                           item_name  \
0         1         1           chips, fresh tomato salsa   
1         1         1                                izze   
2         1         1                    nantucket nectar   
3         1         1  chips, tomatillo green chili salsa   
4         2         2                        chicken bowl   

                                  choice_description  item_price  total_price  
0                                      Not Specified        2.39         2.39  
1                                       [Clementine]        3.39         3.39  
2                                            [Apple]        3.39         3.39  
3                                      Not Specified        2.39         2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98        33.96

# Handling Categorical data

One hot encoding

In [ ]:
df = data

In [ ]:
df_encoded = pd.get_dummies(df, columns=['item_name'])
df_encoded

order_id  quantity                                 choice_description  \
0            1         1                                      Not Specified   
1            1         1                                       [Clementine]   
2            1         1                                            [Apple]   
3            1         1                                      Not Specified   
4            2         2  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   
...        ...       ...                                                ...   
4617      1833         1  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...   
4618      1833         1  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...   
4619      1834         1  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...   
4620      1834         1  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...   
4621      1834         1  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...   

      item_price  total_price  item_name_6 pack soft drink  \
0           2.39         2.39                            0   
1           3.39         3.39                            0   
2           3.39         3.39                            0   
3           2.39         2.39                            0   
4          16.98        33.96                            0   
...          ...          ...                          ...   
4617       11.75        11.75                            0   
4618       11.75        11.75                            0   
4619       11.25        11.25                            0   
4620        8.75         8.75                            0   
4621        8.75         8.75                            0   

      item_name_barbacoa bowl  item_name_barbacoa burrito  \
0                           0                           0   
1                           0                           0   
2                           0                           0   
3                           0                           0   
4                           0                           0   
...                       ...                         ...   
4617                        0                           0   
4618                        0                           0   
4619                        0                           0   
4620                        0                           0   
4621                        0                           0   

      item_name_barbacoa crispy tacos  item_name_barbacoa salad bowl  ...  \
0                                   0                              0  ...   
1                                   0                              0  ...   
2                                   0                              0  ...   
3                                   0                              0  ...   
4                                   0                              0  ...   
...                               ...                            ...  ...   
4617                                0                              0  ...   
4618                                0                              0  ...   
4619                                0                              0  ...   
4620                                0                              0  ...   
4621                                0                              0  ...   

      item_name_steak crispy tacos  item_name_steak salad  \
0                                0                      0   
1                                0                      0   
2                                0                      0   
3                                0                      0   
4                                0                      0   
...                            ...                    ...   
4617                             0                      0   
4618                             0                      0   
4619                             0                      0   
4620                             0          

In [ ]:
data.columns

Index(['order_id', 'quantity', 'item_name', 'choice_description', 'item_price',
       'total_price'],
      dtype='object')

1.	Quantity of each item

In [ ]:
item_quantity = data.groupby('item_name')['quantity'].sum().sort_values(ascending=False)
item_quantity.head()

item_name
chicken bowl         761
chicken burrito      591
chips, guacamole     506
steak burrito        386
canned soft drink    351
Name: quantity, dtype: int64

2.	What was the most ordered item in the choice_description column?

In [ ]:
most_ordered_choice = df.groupby('choice_description')['quantity'].sum().idxmax()
most_ordered_choice

'Not Specified'

3.	How many items were ordered in total?

In [ ]:
total_items_ordered = df['quantity'].sum()
print(total_items_ordered)

4972


4.	How much was the revenue for the period in the dataset?

In [ ]:
total_revenue = (df['quantity'] * df['item_price']).sum()
print(total_revenue)

39237.02


5.	How many orders were made in the period?

In [ ]:
total_orders = df['order_id'].nunique()
print(total_orders)

1834


6.	What is the average revenue amount per order?

In [ ]:
average_revenue_per_order = total_revenue / total_orders
print(average_revenue_per_order)

21.39423118865867


7.	How many different items are sold?

In [ ]:
unique_items_sold = df['item_name'].nunique()
print(unique_items_sold)

47
